lib/src/byte_stream_reader.dart

```diff
@@ -0,0 +1,273 @@
+  Future<int> readInt32([Endian endian]) =>
+      read(4, endian: endian ?? this.endian).then((b) => b.readInt32());
+
+  /// Read a signed, 64-bit integer.
+  ///
+  /// You may provide an [endian] value to override the property set on this instance (See [ByteStreamReader].endian).
+  Future<int> readInt64([Endian endian]) =>
+      read(8, endian: endian ?? this.endian).then((b) => b.readInt64());
+
+  /// Read an unsigned, 8-bit integer.
+  Future<int> readUint8() => read(1).then((b) => b.readUint8());
+
+  /// Read an unsigned, 16-bit integer.
+  ///
+  /// You may provide an [endian] value to override the property set on this instance (See [ByteStreamReader].endian).
+  Future<int> readUint16([Endian endian]) =>
+      read(2, endian: endian ?? this.endian).then((b) => b.readUint16());
+
+  /// Read an unsigned, 32-bit integer.
+  ///
+  /// You may provide an [endian] value to override the property set on this instance (See [ByteStreamReader].endian).
+  Future<int> readUint32([Endian endian]) =>
+      read(4, endian: endian ?? this.endian).then((b) => b.readUint32());
+
+  /// Read an unsigned, 64-bit integer.
+  ///
+  /// You may provide an [endian] value to override the property set on this instance (See [ByteStreamReader].endian).
+  Future<int> readUint64([Endian endian]) =>
+      read(8, endian: endian ?? this.endian).then((b) => b.readUint64());
+
+  /// Read a 32-bit, floating-point number.
+  ///
+  /// You may provide an [endian] value to override the property set on this instance (See [ByteStreamReader].endian).
+  Future<double> readFloat32([Endian endian]) =>
+      read(4, endian: endian ?? this.endian).then((b) => b.readFloat32());
+
+  /// Read a 64-bit, floating-point number.
+  ///
+  /// You may provide an [endian] value to override the property set on this instance (See [ByteStreamReader].endian).
+  Future<double> readFloat64([Endian endian]) =>
+      read(8, endian: endian ?? this.endian).then((b) => b.readFloat64());
+
+  /// Consumes a number of bytes from the input stream (see [consume]), then returns
+  /// a single [ByteDataReader] that reads the received buffer.
+  ///
+  /// You may provide an [endian] value to override the property set on this instance (See [ByteStreamReader].endian).
+  Future<ByteDataReader> read(int length, {bool copy: false, Endian endian}) {
+    return consume(length).then((buf) =>
+        new ByteDataReader(endian: endian ?? this.endian, copy: copy)
+          ..add(buf));
+  }
+
+  /// Consumes a number of bytes from the input stream (see [consume]), then adds the read data
+  /// into a [ByteDataReader].
+  Future<void> readInto(ByteDataReader reader, int length, {bool copy}) {
+    return consume(length).then((buf) => reader.add(buf, copy: copy));
+  }
+
+  /// Returns a [Future] that completes with a buffer of the given size, read
+  /// asynchronously from the incoming stream.
+  Future<Uint8List> consume(int length) {
```

Comment:
This seems to be really costly, build a new list for each chunks of bytes.
`ByteDataReader` already has the capability of accumulating bytes the async version of it could very well reuse that. The async part may create a completer that gets completed when the desired number of bytes is available (or return immediately if they already are). E.g.

````
  final _reader = new ByteDataReader();
  Completer _completer;
  int _lengthRequired;

  void add(List<int> bytes) {
    _reader.add(bytes);
    if (_completer != null && _reader.remainingLength >= _lengthRequired) {
      _completer.complete();
    }
  }

  Future readAhead(int length) async {
    if (_reader.remainingLength >= length) return;
    _lengthRequired = length;
    _completer = new Completer();
    await _completer.future;
    _completer = null;
  }
````

I'm also not sure if it is worth to do all reads in an async fashion, e.g. pushing all individual int parsing to a next event loop, or would it be better to just wait for n bytes each time and do the parsing synchronously?

Reply:
Because, if the later is the case, this could be an extension of the `ByteDataReader` class, mixed in with the `StreamConsumer` class.

Reply:
I've updated `ByteDataReader` with `readAhead`. Could you please check how that would make your case simpler?

Reply:
That's perfect TBH. I'm going to go ahead and close this PR, as it's no longer necessary.